### PR TITLE
Add typecheck tests for dataclass attributes

### PR DESCRIPTION
**Pull request by david-yz-liu**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :memo: Test update (I made up a new type, I didn't know where this fit) |

## Description

Added additional tests for some basic type-checking of attributes from dataclasses, motivated by the changes in PyCQA/astroid#1126.

### Comments

I didn't add an entry to the Changelog, but I can if you want.

I identified a few false negatives as well in some of the typecheck checks, but I didn't submit any code changes here.

Also, wow the `invalid-metaclass` error output is... not great?

### CODE REVIEW

File: tests/functional/d/deprecated/dataclass_typecheck.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
...mag:\n        """"""the imaginary part of a complex number""""""\n        \'imag\'\n        \'imag\'\n    \n    \n    \n    class numerator:\n        """"""the numerator of a rational number in lowest terms""""""\n        \'numerator\'\n        \'numerator\'\n    \n    \n    \n    class real:\n        """"""the real part of a complex number""""""\n        \'real\'\n        \'real\'\n    \n    \n    def to_bytes():\n        """"""Return an array of bytes representing an integer.\n\n  length\n    Length of bytes object to use.  An OverflowError is raised if the\n    integer is not representable with the given number of bytes.\n  byteorder\n    The byte order used to represent the integer.  If byteorder is \'big\',\n    the most significant byte is at the beginning of the byte array.  If\n    byteorder is \'little\', the most significant byte is at the end of the\n    byte array.  To request the native byte order of the host system, use\n    `sys.byteorder\' as the byte order value.\n  signed\n    Determines whether two\'s complement is used to represent the integer.\n    If signed is False and a negative integer is given, an OverflowError\n    is raised.""""""\n        \n' used":HIGH
```
Comment (Pierre-Sassoulas) [inline tests/functional/d/deprecated/dataclass_typecheck.txt:10]: Wow. Do you know how we could we fix that ? I know this is not really the goal of this MR but this is... bad.

Comment (david-yz-liu) [reply tests/functional/d/deprecated/dataclass_typecheck.txt:10]: I'm guessing an `as_string` is being called on a node, but it's not immediately clear to me why. I can look into this in a bit, let me know if you want me to add to this PR or open up a new one.

**Comment by david-yz-liu** [comment]
Update: I had to update the test astroid version to 2.7.1 (to include the patch in PyCQA/astroid#1130). Is this okay? Should I have updated `setup.cfg` as well?

**Comment by Pierre-Sassoulas** [comment]
It's okay to increase the astroid version, actually I forgot to do it after releasing 2.7.1 so thank you. But yes both versions should be in sync, or the test can run properly but the installation fail if the wrong version of astroid is used.

File: tests/functional/d/deprecated/dataclass_typecheck.txt (hunk 1 above)
Comment (david-yz-liu) [reply tests/functional/d/deprecated/dataclass_typecheck.txt]: @Pierre-Sassoulas okay never mind, I fixed the immediate problem in this MR. I think the `builtins.int` class that astroid defines has different methods depending on Python version, which caused the test output to be inconsistent.

Comment (Pierre-Sassoulas) [reply tests/functional/d/deprecated/dataclass_typecheck.txt]: :+1: I think it's a mistake, there's no way we have such a messy warning on purpose :smile: Could you fix this in another MR then rebase this one on main when we've merged it ? :)

**Comment by Pierre-Sassoulas** [comment]
Thank you for that and the prior refactor !

**Pull request merged**